File: important/TwoSum.java
class TwoSum {

  // O(n) space and time
  public int[] twoSum(int[] nums, int target) {
    Map<Integer, Integer> numToIndex = new HashMap<>();
    for (int i = 0; i < nums.length; i++) {
      int num = nums[i];
      int targetRemain = target - num;
      if (numToIndex.containsKey(targetRemain)) {
        return new int[] { i, numToIndex.get(targetRemain) };
      }
      numToIndex.put(num, i);
    }

    return new int[] { -1, -1 };
  }

  // brute force
  // O(n^2) time
  // O(1) space
  public int[] twoSum(int[] nums, int target) {
    // i and j are indexes of nums
    for (int i = 0; i < nums.length; i++) {
      for (int j = i + 1; j < nums.length; j++) {
        if (nums[i] + nums[j] == target) {
          return new int[] { i, j };
        }
      }
    }

    return new int[] { 0, 0 };
  }
}

File: P0/TwoSum.py
from typing import List


# dict O(n) space and time
class TwoSum:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        num_to_index = dict()
        for i, num in enumerate(nums):
            target_remain = target - num
            if target_remain in num_to_index:
                return [i, num_to_index.get(target_remain)]

            num_to_index[num] = i

        return [-1, -1]


# brute force O(n^2) time O(1) space
class TwoSum2:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        length = len(nums)
        # i and j are indexes of nums
        for i in range(0, length):
            for j in range(i + 1, length):
                if nums[i] + nums[j] == target:
                    return [i, j]

        return [-1, -1]
